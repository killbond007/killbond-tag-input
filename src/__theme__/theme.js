const theme = {
	breakpoints: ['768px', '1024px', '1240px'],
	devices: {
		sm: '@media (max-width: 576px)',
		md: '@media (max-width: 768px)',
		lg: '@media (max-width: 1024px)',
		xl: '@media (max-width: 1240px)',
	},
	colors: {
		purple_1: '#3827b0',
		purple_2: '#834cc4',
		pale_purple1: '#f9f9ff',
		red: '#ff7068',
		green: '#0cdcae',
		dark_indigo: '#180e5e',
		grey_shades: '#9f9bac',
		light_grey: '#c5d0de',
		dark: '#352c4b',
		white: '#ffffff',
	},
	opacities: {
		opacity_1: 0.85,
		opacity_2: 0.7,
		opacity_3: 0.55,
		opacity_4: 0.4,
		opacity_5: 0.25,
		opacity_6: 0.1,
	},
	fonts: {
		primary: 'Ubuntu',
		icons: 'Otto',
	},
	typography: {
		h1: {
			fontFamily: 'Ubuntu, sans-serif',
			fontSize: 24,
			fontWeight: 500,
			fontStyle: 'normal',
			lineHeight: 1.25,
			color: '#3827b0',
			'@media (max-width: 768px)': {
				fontSize: 22,
				lineHeight: 1.32,
			},
		},
		h2: {
			fontFamily: 'Ubuntu, sans-serif',
			fontSize: 18,
			fontWeight: 'normal',
			fontStyle: 'normal',
			lineHeight: 1.39,
			color: '#180e5e',
			'@media (max-width: 768px)': {
				fontSize: 17,
				lineHeight: 1.35,
			},
		},
		h3: {
			fontFamily: 'Ubuntu, sans-serif',
			fontSize: 16,
			fontWeight: 500,
			fontStyle: 'normal',
			lineHeight: 1.25,
			color: '#180e5e',
			'@media (max-width: 768px)': {
				fontSize: 15,
				lineHeight: 1.47,
			},
		},
		h4: {
			fontFamily: 'Ubuntu, sans-serif',
			fontSize: 15,
			fontWeight: 500,
			fontStyle: 'normal',
			lineHeight: 1.33,
			color: '#180e5e',
			'@media (max-width: 768px)': {
				fontSize: 14,
				lineHeight: 1.33,
			},
		},
		body: {
			fontFamily: 'Ubuntu, sans-serif',
			fontSize: 14,
			fontWeight: 'normal',
			fontStyle: 'normal',
			lineHeight: 1.43,
			color: '#352c4b',
			'@media (max-width: 768px)': {
				fontSize: 13,
				lineHeight: 1.54,
			},
		},
		body_bold: {
			fontFamily: 'Ubuntu, sans-serif',
			fontSize: 14,
			fontWeight: 700,
			fontStyle: 'normal',
			lineHeight: 1.43,
			color: '#352c4b',
			'@media (max-width: 768px)': {
				fontSize: 13,
				lineHeight: 1.54,
			},
		},
		label: {
			fontFamily: 'Ubuntu, sans-serif',
			fontSize: 12,
			fontWeight: 'normal',
			fontStyle: 'normal',
			lineHeight: 1.33,
			color: '#3827b0',
			'@media (max-width: 768px)': {
				fontSize: 11,
				lineHeight: 1.27,
			},
		},
		label_form: {
			fontFamily: 'Ubuntu, sans-serif',
			fontSize: 12,
			fontWeight: 'normal',
			fontStyle: 'normal',
			lineHeight: 1.33,
			color: '#9f9bac',
			'@media (max-width: 768px)': {
				fontSize: 11,
				lineHeight: 1.27,
			},
		},
		label_link: {
			fontFamily: 'Ubuntu, sans-serif',
			fontSize: 12,
			fontWeight: 'normal',
			fontStyle: 'normal',
			lineHeight: 1.33,
			color: '#9f9bac',
			'@media (max-width: 768px)': {
				fontSize: 11,
				lineHeight: 1.27,
			},
		},
		th: {
			fontFamily: 'Ubuntu, sans-serif',
			textTransform: 'uppercase',
			fontSize: 9,
			fontWeight: 'normal',
			fontStyle: 'normal',
			lineHeight: 1.33,
			color: '#9f9bac',
		},
		figure: {
			fontFamily: 'Ubuntu, sans-serif',
			fontSize: 40,
			fontWeight: 500,
			fontStyle: 'normal',
			lineHeight: 1.13,
			color: '#352c4b',
			'@media (max-width: 768px)': {
				fontSize: 30,
			},
		},
		icon_small: {
			'&&&': {
				fontFamily: 'Otto',
			},
			fontSize: 15,
			fontWeight: 'normal',
			fontStyle: 'normal',
			lineHeight: 1,
			'-webkit-font-smoothing': 'antialiased',
			'-moz-osx-font-smoothing': 'grayscale',
			color: '#ffffff',
		},
		icon_large: {
			'&&&': {
				fontFamily: 'Otto',
			},
			fontSize: 24,
			fontWeight: 'normal',
			fontStyle: 'normal',
			lineHeight: 1,
			'-webkit-font-smoothing': 'antialiased',
			'-moz-osx-font-smoothing': 'grayscale',
			color: '#ffffff',
		},
		icon_logo: {
			'&&&': {
				fontFamily: 'Otto',
			},
			fontSize: 35,
			fontWeight: 'normal',
			fontStyle: 'normal',
			lineHeight: 1,
			'-webkit-font-smoothing': 'antialiased',
			'-moz-osx-font-smoothing': 'grayscale',
			color: '#ffffff',
		},
	},
	shadows: {
		small: {
			boxShadow: '1px 1px 0 0 rgba(197, 208, 222, 0.4)',
		},
		medium: {
			boxShadow: '0 2px 1px 0 rgba(197, 208, 222, 0.7)',
		},
		large: {
			boxShadow: '0 0 2px 0 rgba(197, 208, 222, 0.7)',
		},
		table: {
			boxShadow: '4px 4px 4px 0 rgba(197, 208, 222, 0.4)',
		},
		focused: {
			boxShadow: '0 0 2px 0 #834cc4',
		},
	},
	borders: {
		light: {
			borderRadius: 3,
			border: '1px rgba(249, 249, 255, 1) solid',
		},
		regular: {
			borderRadius: 3,
			border: '1px rgba(197, 208, 222, 0.4) solid',
		},
		bold: {
			borderRadius: 3,
			border: '1px #c5d0de solid',
		},
		focused: {
			borderRadius: 3,
			border: '1px #834cc4 solid',
		},
		selected: {
			border: '4px #0cdcae solid',
		},
		error: {
			borderRadius: 3,
			border: '1px #ff7068 solid',
		},
	},
	lines: {
		separator: {},
	},
}

export default theme
