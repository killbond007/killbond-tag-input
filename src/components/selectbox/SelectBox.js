import React, { Fragment, useEffect, useRef, useState } from 'react'
import PropTypes from 'prop-types'
import { noop } from 'lodash'

import Tag from '../tag/Tag'

import * as Styled from './__styles__/SelectBox.styles'

const SelectBox = ({
	children,
	id,
	name,
	value,
	width,
	isMulti,
	isSearchable,
	isDisabled,
	placeholder,
	emptyLabel,
	autoFocus,
	defaultOpen,
	onSelect,
	...restProps
}) => {
	const inputRef = useRef(null)
	const rootRef = useRef(null)

	const [isOpen, setIsOpen] = useState(defaultOpen)
	const [inputValue, setInputValue] = useState('')
	const [inputWidth, setInputWidth] = useState(0)

	useEffect(() => {
		document.addEventListener('click', onClickOutside, false)
		return () => {
			document.removeEventListener('click', onClickOutside, false)
		}
	}, [isOpen])
	const onClickOutside = (e) => {
		if (rootRef.current && !rootRef.current.contains(e.target)) {
			setIsOpen(false)
		}
	}

	const toggle = () => {
		setIsOpen((isOpen) => !isOpen)
	}

	const onSelectBoxClick = (e) => {
		e.target.getAttribute('name') !== 'close' && toggle()
		isSearchable && inputRef.current.focus()
	}

	const onSelectOptionClick = (_value) => {
		if (isMulti) {
			value
				? Array.isArray(value)
					? onSelect(name, [...value, _value], id)
					: onSelect(name, [value, _value], id)
				: onSelect(name, [_value], id)
		} else {
			onSelect(name, _value, id)
		}
		if (isSearchable) {
			setInputValue('')
			setInputWidth(0)
			inputRef.current.value = ''
		}
		setIsOpen(false)
	}

	const onTagDelete = (i) => {
		const newValue = Array.isArray(value) ? [...value] : [value]
		newValue.splice(i, 1)
		onSelect(name, newValue, id)
	}

	const onInputChange = (e) => {
		const { value: inputValue, scrollWidth: inputWidth } = e.target
		setIsOpen(true)
		setInputValue(inputValue)
		setInputWidth(inputWidth)
	}

	const onInputKeyDown = (e) => {
		const { value: inputValue } = e.target
		if (!inputValue && e.key === 'Backspace') {
			value && onTagDelete(value.length - 1)
		}
	}

	let multiOptions = []

	let items = React.Children.toArray(children)

	if (items && isMulti && value) {
		multiOptions = items.filter(
			(child) =>
				value.indexOf(child.props.value) === -1 &&
				child.props.label.toLowerCase().indexOf(inputValue.toLowerCase()) !== -1
		)
	}

	let selectedValues = isMulti && value ? (Array.isArray(value) ? value : [value]) : []

	return (
		<Styled.Root ref={rootRef} {...restProps}>
			<Styled.Box
				data-testid="__select-box__"
				width={width}
				isDisabled={isDisabled}
				isOpen={isOpen}
				isMulti={isMulti}
				autoFocus={autoFocus}
				onClick={!isDisabled ? onSelectBoxClick : null}
			>
				{isMulti ? (
					<Styled.MultiSelection>
						{selectedValues.map(
							(item, i) =>
								items.find(({ props }) => props.value === item) && (
									<Styled.Tag key={`value-${i}`}>
										<Tag id={i} onDelete={onTagDelete} hex={items.find(({ props }) => props.value === item).props.hex}>
											{items.find(({ props }) => props.value === item).props.label}
										</Tag>
									</Styled.Tag>
								)
						)}
						{isSearchable ? (
							<Styled.Input
								ref={inputRef}
								placeholder={selectedValues.length === 0 ? placeholder : null}
								width={selectedValues.length === 0 ? '100%' : inputWidth}
								onChange={onInputChange}
								onKeyDown={onInputKeyDown}
							/>
						) : (
							selectedValues.length === 0 && <Styled.Placeholder>{placeholder}</Styled.Placeholder>
						)}
					</Styled.MultiSelection>
				) : isSearchable ? (
					<Styled.Selection>
						{items.find(({ props }) => props.value === value && props.icon !== null) && (
							<Styled.Icon
								data-testid="__select-icon__"
								isOpen={isOpen}
								name={items.find(({ props }) => props.value === value).props.icon}
							/>
						)}
						<Styled.SingleInput
							ref={inputRef}
							placeholder={value ? items.find(({ props }) => props.value === value).props.label : placeholder}
							width={width - 40}
							isSelected={value}
							onChange={onInputChange}
						/>
					</Styled.Selection>
				) : !value ? (
					<Styled.SelectPlaceholder>{placeholder}</Styled.SelectPlaceholder>
				) : (
					<Styled.Selection>
						{items.find(({ props }) => props.value === value && props.icon !== null) && (
							<Styled.Icon
								data-testid="__select-icon__"
								isOpen={isOpen}
								name={items.find(({ props }) => props.value === value)?.props?.icon}
							/>
						)}
						<Styled.Label>{items.find(({ props }) => props.value === value)?.props?.label}</Styled.Label>
					</Styled.Selection>
				)}
				<Styled.Arrow isOpen={isOpen} />
			</Styled.Box>
			{isOpen && (
				<Fragment>
					<Styled.Menu>
						{isMulti
							? value &&
							  (items.length === value.length || multiOptions.length === 0) && (
									<Styled.HasNoOptions>{emptyLabel}</Styled.HasNoOptions>
							  )
							: items.filter((child) => child.props.label.toLowerCase().indexOf(inputValue.toLowerCase()) !== -1)
									.length === 0 && <Styled.HasNoOptions>{emptyLabel}</Styled.HasNoOptions>}
						{items &&
							React.Children.map(items, (child, i) =>
								isMulti
									? (value ? value.indexOf(child.props.value) === -1 : true) &&
									  child.props.label.toLowerCase().indexOf(inputValue.toLowerCase()) !== -1 &&
									  React.cloneElement(child, {
											key: `option-${i}`,
											onClick: onSelectOptionClick,
									  })
									: isSearchable
									? child.props.label.toLowerCase().indexOf(inputValue.toLowerCase()) !== -1 &&
									  React.cloneElement(child, {
											key: `option-${i}`,
											onClick: onSelectOptionClick,
									  })
									: React.cloneElement(child, {
											key: `option-${i}`,
											onClick: onSelectOptionClick,
									  })
							)}
					</Styled.Menu>
					<Styled.Line />
				</Fragment>
			)}
		</Styled.Root>
	)
}

SelectBox.propTypes = {
	/** Defines the id of the selectBox */
	id: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
	/** Defines the name of the selectBox */
	name: PropTypes.string,
	/** Defines the value of the selectBox */
	value: PropTypes.oneOfType([PropTypes.any, PropTypes.arrayOf(PropTypes.any)]),
	/** Defines the width of the selectBox */
	width: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
	/** Defines if several items can be selected in the selectBox */
	isMulti: PropTypes.bool,
	/** Defines if the items can be searched in the selectBox */
	isSearchable: PropTypes.bool,
	/** Defines the default label when no selected item */
	placeholder: PropTypes.string,
	/** Defines the text display in the empty menu */
	emptyLabel: PropTypes.string,
	/** Defines the default focus by default */
	autoFocus: PropTypes.bool,
	/** Defines the default open state of dropdown */
	defaultOpen: PropTypes.bool,
	/** Defines the handler called when an option is selected */
	onSelect: PropTypes.func,
	/** Defines the options type */
	children: PropTypes.oneOfType([PropTypes.node, PropTypes.arrayOf(PropTypes.node)]),
}

SelectBox.defaultProps = {
	id: null,
	name: null,
	value: null,
	width: 'auto',
	isMulti: false,
	isSearchable: false,
	placeholder: null,
	emptyLabel: null,
	autoFocus: false,
	defaultOpen: false,
	onSelect: noop,
	children: null,
}

export default SelectBox
