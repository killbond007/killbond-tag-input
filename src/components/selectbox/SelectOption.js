import React from 'react'
import PropTypes from 'prop-types'
import { noop } from 'lodash'

import * as Styled from './__styles__/SelectOption.styles'

const SelectOption = ({ value, label, onClick, icon }) => {
	const _onClick = () => onClick(value)

	return (
		<Styled.Root data-testid="__select-option__" value={value} onClick={_onClick}>
			{icon && <Styled.Icon data-testid="__select-option-icon__" name={icon} />}
			<Styled.Label>{label}</Styled.Label>
		</Styled.Root>
	)
}

SelectOption.propTypes = {
	/** Defines the value of the option */
	value: PropTypes.any,
	/** Defines the label of the option */
	label: PropTypes.string,
	/** Defines the source of the icon  */
	icon: PropTypes.string,
	/** Defines the handler called on click */
	onClick: PropTypes.func,
}

SelectOption.defaultProps = {
	value: null,
	label: null,
	icon: null,
	onClick: noop,
}

export default SelectOption
