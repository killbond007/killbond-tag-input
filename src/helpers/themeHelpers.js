import Color from 'color'
import { get } from 'lodash'

export const getColor = (theme, color = '#000', opacity = 1) => {
	color = get(theme, `colors[${color}]`) || color
	opacity = get(theme, `opacities[${opacity}]`) || opacity
	return Color(color).alpha(opacity).string()
}

export const getColorJest = (...args) => {
	return getColor(...args).replace(/, /g, ',')
}
