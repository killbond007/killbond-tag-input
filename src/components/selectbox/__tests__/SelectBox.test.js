import React from 'react'
import { fireEvent, render } from '@testing-library/react'
import { ThemeProvider } from 'styled-components'

import SelectBox from '../SelectBox'
import SelectOption from '../SelectOption'

import { getColorJest } from '../../../helpers/themeHelpers'

import theme from '../../../__theme__/theme'

const defaultProps = {}
const getInstance = (props = {}) => (
	<ThemeProvider theme={theme}>
		<SelectBox {...defaultProps} {...props}>
			<SelectOption value="foo" label="Foo" />
			<SelectOption value="bar" label="Bar" />
			<SelectOption value="gag" label="Gag" />
		</SelectBox>
	</ThemeProvider>
)

describe('SelectBox', () => {
	it('matches snapshot', () => {
		const { asFragment } = render(getInstance())
		expect(asFragment()).toMatchSnapshot()
	})

	it('renders width', () => {
		const width = 200
		const { getByTestId } = render(getInstance({ width }))
		expect(getByTestId('__select-box__')).toHaveStyleRule('width', `${width}px`)
	})

	it('renders placeholder', () => {
		const placeholder = 'Select an option'
		const { getByText } = render(getInstance({ placeholder }))
		expect(getByText(placeholder)).toBeInTheDocument()
	})

	it('renders value', () => {
		const value = 'foo'
		const { getByText } = render(getInstance({ value }))
		expect(getByText('Foo')).toBeInTheDocument()
	})

	it('renders isDisabled', () => {
		const { getByTestId } = render(getInstance({ isDisabled: true }))
		expect(getByTestId('__select-box__')).toHaveStyleRule(
			'background-color',
			getColorJest(theme, 'light_grey', 'opacity_5')
		)
	})

	it('triggers onSelect callback', () => {
		const onSelect = jest.fn()
		const { getAllByTestId, getByTestId } = render(getInstance({ onSelect }))
		fireEvent.click(getByTestId('__select-box__'))
		fireEvent.click(getAllByTestId('__select-option__')[0])
		expect(onSelect).toHaveBeenCalledWith(null, 'foo', null)
	})

	it('renders options on click', () => {
		const { queryAllByTestId, getByTestId } = render(getInstance())
		expect(queryAllByTestId('__select-option__')).toHaveLength(0)
		fireEvent.click(getByTestId('__select-box__'))
		expect(queryAllByTestId('__select-option__')).toHaveLength(3)
	})
})
