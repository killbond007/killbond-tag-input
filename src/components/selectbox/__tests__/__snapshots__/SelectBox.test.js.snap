// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectBox matches snapshot 1`] = `
<DocumentFragment>
  .c0 {
  position: relative;
  display: inline-block;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c1 {
  font-family: Ubuntu,sans-serif;
  font-size: 14px;
  font-weight: normal;
  font-style: normal;
  line-height: 1.43;
  color: rgb(56,39,176);
  box-shadow: 1px 1px 0 0 rgba(197,208,222,0.4);
  width: auto;
  border-radius: 3px;
  min-height: 40px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  padding: 4px 16px;
  text-align: left;
  background-color: rgb(255,255,255);
  border: 1px solid rgb(197,208,222);
  border-bottom-color: rgb(197,208,222);
}

.c1:focus {
  border-radius: 3px;
  border: 1px #834cc4 solid;
}

.c3 {
  width: 0;
  height: 0;
  border: 5px solid transparent;
  margin-top: 6px;
  margin-bottom: 0;
  border-color: rgb(56,39,176) transparent transparent transparent;
}

.c2 {
  font-family: Ubuntu,sans-serif;
  font-size: 14px;
  font-weight: normal;
  font-style: normal;
  line-height: 1.43;
  color: rgb(53,44,75);
}

@media (max-width:768px) {
  .c1 {
    font-size: 13px;
    line-height: 1.54;
  }
}

@media (max-width:768px) {
  .c2 {
    font-size: 13px;
    line-height: 1.54;
  }
}

<div
    class="c0"
  >
    <button
      class="c1"
      data-testid="__select-box__"
      width="auto"
    >
      <span
        class="c2"
      />
      <div
        class="c3"
      />
    </button>
  </div>
</DocumentFragment>
`;
